typegen/unification/subtyping.py

```diff
@@ -0,0 +1,156 @@
+import importlib
+import importlib.util
+from importlib.machinery import SourceFileLoader
+import logging
+import os
+from types import ModuleType
+import pandas as pd
+import pathlib
+import sys
+
+from tracing.resolver import Resolver
+
+from .filter_base import TraceDataFilter
+import constants
+
+logger = logging.getLogger(__name__)
+
+
+def _none_if_builtin(module: str) -> str | None:
+    return module if module != "builtins" else None
+
+
+class ReplaceSubTypesFilter(TraceDataFilter):
+    """Replaces rows containing types in the data with their common base type."""
+
+    ident = "repl_subty"
+
+    stdlib_path: pathlib.Path | None = None
+    proj_path: pathlib.Path | None = None
+    venv_path: pathlib.Path | None = None
+    only_replace_if_base_was_traced: bool | None = None
+
+    def apply(self, trace_data: pd.DataFrame) -> pd.DataFrame:
+        """
+        Replaces the rows containing types with their common base type and returns the processed trace data. If
+        only_replace_if_base_type_already_in_data is True, only rows of types whose base type is already in the data
+        are replaced.
+
+        @param trace_data The provided trace data to process.
+        """
+        if self.stdlib_path is None:
+            raise AttributeError(
+                f"{ReplaceSubTypesFilter.__name__} was not initialised properly: {self.stdlib_path=}"
+            )
+        if self.proj_path is None:
+            raise AttributeError(
+                f"{ReplaceSubTypesFilter.__name__} was not initialised properly: {self.proj_path=}"
+            )
+        if self.venv_path is None:
+            raise AttributeError(
+                f"{ReplaceSubTypesFilter.__name__} was not initialised properly: {self.venv_path=}"
+            )
+        if self.only_replace_if_base_was_traced is None:
+            raise AttributeError(
+                f"{ReplaceSubTypesFilter.__name__} was not initialised properly: {self.only_replace_if_base_was_traced=}"
+            )
+
+        self._resolver = Resolver(self.stdlib_path, self.proj_path, self.venv_path)
```

Comment:
I recommend instantiating the resolver outside the class and giving it as a reference. This would be useful as the resolver is always the same and can be used for other classes (for example, InlineGenerator)